### PR TITLE
A2-1488 Support attachments in ITSM backend reference implementation

### PR DESCRIPTION
**Pull request by Wade-GitHub**
Attachments can be posted as base64 data in a field inside the request body.

Request body would look something like:

```json
{
  "ticket_id": 3,
  "filename": "text_to_post.txt",
  "content_type": "text/plain",
  "attachment_data": "VGhpcyBpcyBhIGZpbGUgdG8gY2hlY2sgdGhhdCBwb3N0aW5nIGEgbmV3IGF0dGFjaG1lbnQgd29ya3MK"
}
```

### CODE REVIEW

File: test_itsm.py [hunk 1]
```diff
@@ -815,3 +817,208 @@ def test_create_comment(client):
...
+    # Get the url to the attachments list resource, then get the list of attachments that are
+    # currently stored in the database.
+    attachments_url = _get_root_links(client)['attachments']
+    attachments     = client.get(attachments_url, **JSON_HDRS_READ).get_json()
+
+    # The test jpeg image attachment's base64 encoding is quite large - so here we use the
+    # base64 library to create the encoding and store it in a variable instead of having the
+    # literal string in the assert data.
+    with open("test_data/mt-fuji.jpeg", "rb") as image_attachment:
+        encoded_image_attachment = base64.b64encode(image_attachment.read()).decode()
```
Comment (jbrendel) [inline test_itsm.py]: It seems this variable isn't used in this test?

Comment (Wade-GitHub) [reply test_itsm.py]: Right, an artifact from some previous testing. Removed.

File: itsm_api/views.py [hunk 2]
```diff
@@ -833,48 +833,54 @@ def _post(self, data):
...
+        # The directory structure of saved attachment files looks like this:
+        #
+        # attachment_storage/
+        #   <ticket_id>/
+        #     <unique_filename>
+        #
+        # For example:
+        #
+        # attachment_storage/
+        #   1/
```
Comment (jbrendel) [inline itsm_api/views.py:853]: I think this docstring needs to be updated, now that the directories are called `ticket__<id>` ?

Comment (Wade-GitHub) [reply itsm_api/views.py]: Oh yes, updated to reflect proper directory storage

File: itsm_api/views.py [hunk 3]
```diff
@@ -806,6 +791,120 @@ def _post(self, data):
         return new_comment_id
 
 
+class AttachmentList(flask_restful.Resource, ApiResourceList):
+    """
+    A list of attachments.
+
+    Similar to the comment list, this resource is not 'user friendly' to the
+    extend that it does not contain embedded resources. Instead, it is merely
```
Comment (natchu) [inline itsm_api/views.py:799]: `extent`? or `to extend`?

File: test_itsm.py [hunk 4]
```diff
@@ -815,3 +813,201 @@ def test_create_comment(client):
 
     # Should be the same as what we got in our POST response
     assert post_response == comment
+
+
+def test_attachments_list(client):
```
Comment (natchu) [inline test_itsm.py:818]: test_get_attachments_list()

File: test_itsm.py [hunk 5]
```diff
@@ -815,3 +813,201 @@ def test_create_comment(client):
...
+            },
+            "contained_in": {
+                "href": "/"
+            }
+        }
+    }
+
+
+def test_get_attachment(client):
+    # First check we get the proper response when trying to get an attachment that doesn't
```
Comment (natchu) [inline test_itsm.py:865]: `if we get`

File: test_itsm.py [hunk 6]
```diff
@@ -815,3 +813,201 @@ def test_create_comment(client):
...
+    # exist
+    no_attachment_url = _get_root_links(client)['attachments'] + "/999"
+    no_attach_resp    = client.get(no_attachment_url, **JSON_HDRS_READ)
+    assert no_attach_resp.is_json and no_attach_resp.status_code == 404
+
+    resp_message = no_attach_resp.get_json()['message']
```
Comment (natchu) [inline test_itsm.py:871]: align

File: itsm_api/views.py (hunk 3 above)
Comment (Wade-GitHub) [reply itsm_api/views.py]: Should be `extent`, updated

File: test_itsm.py (hunk 4 above)
Comment (Wade-GitHub) [reply test_itsm.py]: Updated

File: test_itsm.py (hunk 5 above)
Comment (Wade-GitHub) [reply test_itsm.py]: Updated

File: itsm_api/views.py [hunk 7]
```diff
@@ -806,6 +791,120 @@ def _post(self, data):
...
+            DB_ATTACHMENT_TABLE.remove(doc_ids=[new_attachment_id])
+            flask_restful.abort(500, message="Error occured while trying to save attachment "
+                                             "file data")
+        except Exception:
+            # Some error occured while trying to decode the attachment file data. Rollback the
+            # database entry and return a 500 error.
+            DB_ATTACHMENT_TABLE.remove(doc_ids=[new_attachment_id])
+            flask_restful.abort(500, message="Error occured while trying to decode "
+                                             "attachment file data")
+        else:
```
Comment (jbrendel) [inline itsm_api/views.py:903]: Why do we even need the else branch here? If I understand correctly, `flash_restful.abort()` will end the execution, will it not? Alternatively, we can also remove else and just move line 904 and 905 to be in line 889.

Comment (Wade-GitHub) [reply itsm_api/views.py]: Ah right, I was thinking along the lines of "we only want to return the ID if it works", but I can see that if we're catching `Exception`, then `else` doesn't add any extra value. Updated to remove the else block

**Pull request merged**